### PR TITLE
Microos: Avoid 'Stall detected' problems

### PR DESCRIPTION
**Pull request by jlausuch**
It's not possible to disable grub timeout in RAW image
due to its RO filesystem nature. Sometimes, OpenQA
takes more than 10 seconds to match a needle and the
grub2 screen is gone and fails with "Stall detected"
error.
This way, we match either grub2 or login screen.

Example VRs:
microos-5.0-MicroOS-Image-A-Staging-x86_64-BuildA.10.10
[bad](https://openqa.suse.de/tests/4836727#step/transactional_update/57) [fixed](http://fromm.arch.suse.de/tests/467)

microos-5.0-DVD-A-Staging-x86_64-BuildA.44.3
[bad](https://openqa.suse.de/tests/4848974#step/transactional_update/29) [fixed](http://fromm.arch.suse.de/tests/468)

Regression in openSUSE:
[microos-15.2-MicroOS-Image-x86_64-Build35.53-microos@64bit_virtio](http://fromm.arch.suse.de/tests/473)
[microos-15.2-DVD-x86_64-Build35.53-microos@64bit_virtio](http://fromm.arch.suse.de/tests/474)
[microos-Tumbleweed-DVD-x86_64-Build20201017-microos@64bit](http://fromm.arch.suse.de/tests/471)
[microos-Tumbleweed-MicroOS-Image-x86_64-Build20201017-microos@64bit](http://fromm.arch.suse.de/tests/472)
NOTE: Errors in Leap are due to something else.

### CODE REVIEW

File: lib/microos.pm [hunk 1]
```diff
@@ -22,7 +22,7 @@ our @EXPORT = qw(microos_reboot microos_login);
 
 # Assert login prompt and login as root
 sub microos_login {
-    assert_screen 'linux-login-microos', 150;
+    assert_screen('linux-login-microos', 150) unless match_has_tag('linux-login-microos');
```
Comment (Vogtinator) [inline lib/microos.pm:25]: Is this change necessary?

This might break if `microos_login` is called after a tty switch or so, but currently that's not the case

**Review by Vogtinator — COMMENTED** [review]
> It's not possible to disable grub timeout in RAW image
due to its RO filesystem nature.

It is, just needs a new snapshot. I agree that it's cleaner to work without needing that though.

AFAICT this is a partial revert of https://github.com/os-autoinst/os-autoinst-distri-opensuse/commit/f2c6575b2e10b45eb0899f385aa3680c6bb35390? Is the change in autoyast/installation.pm not necessary?

File: lib/microos.pm (hunk 1 above)
Comment (jlausuch) [reply lib/microos.pm:25]: I'm trying to work on staging images for SUSE MicroOS, there are some recent changes that need needling. So, I put this in WIP until I find some stability in those images and also will run VRs on openSUSE jobs. Sorry if I added you as reviewer too early.

**Comment by jlausuch** [comment]
More VRs:
[microos-5.0-DVD-x86_64-Build35.1-suse_microos_installation+update@64bit](http://fromm.arch.suse.de/tests/476)
[microos-5.0-MicroOS-Image-A-Staging-x86_64-BuildA.10.10-suse_microos_update@64bit](http://fromm.arch.suse.de/tests/475)
[microos-5.0-MicroOS-Image-x86_64-Build8.1-suse_microos_update@64bit](http://fromm.arch.suse.de/tests/477)

File: lib/microos.pm [hunk 2]
```diff
@@ -22,7 +22,9 @@ our @EXPORT = qw(microos_reboot microos_login);
 
 # Assert login prompt and login as root
 sub microos_login {
-    assert_screen 'linux-login-microos', 150;
+    unless (match_has_tag('linux-login-microos') || match_has_tag('linux-login')) {
```
Comment (Vogtinator) [inline lib/microos.pm:25]: I don't get the differentiation between `linux-login-microos` and `linux-login` in needles, AFAICT it's all the same and there can't really be any conflict. But that's a topic for some other time...

In this particular case I wonder why it's fine with `linux-login` as already matched tag, but doesn't accept it in the `assert_screen` here or in below in line 52.

Comment (jlausuch) [reply lib/microos.pm:25]: The reason for adding `linux-login` in the condition is to avoid matching the same stuff twice and save 1 step. For example [here](https://openqa.suse.de/tests/4834939#step/disk_boot/2)  matches first `linux-login` and in the next frame it matches `linux-login-microos`.

Maybe I can just skip that condition if it's more clear, as it works anyway.

Comment (Vogtinator) [reply lib/microos.pm:25]: It sounds like whatever matched `linux-login` there should also match `linux-login-microos`, then the second `match_has_tag` here could be dropped again and it's consistenly `linux-login-microos` throughout.

Comment (Vogtinator) [reply lib/microos.pm]: AFAICT it's from `wait_boot_textmode`

Comment (jlausuch) [reply lib/microos.pm]: > AFAICT it's from `wait_boot_textmode`
Yes, `disk_boot` calls `wait_boot`, which calls `wait_boot_past_bootloader` and `wait_boot_textmode`. This will match `linux-login` needle and then we are also forcing `linux-login-microos` as well...

See this error if I remove it from the condition as it is now:  http://fromm.arch.suse.de/tests/487#

Comment (Vogtinator) [reply lib/microos.pm]: Yeah, either `linux-login-microos` should also have the `linux-login` tag or every use of `linux-login` should also match `linux-login-microos` there. This particular failure looks like either a product bug (screen not cleared) or a missing needle for the case that `login:` is at the bottom.

Comment (jlausuch) [reply lib/microos.pm]: yeah, btw, the `linux-login-microos` needle already has the tag `linux-login`.

Comment (Vogtinator) [reply lib/microos.pm]: Not here (openSUSE):

```
/var/lib/openqa/tests/opensuse/products/opensuse/needles> git grep linux-login *.json
autoyast-system-login-console-minimal-20181012.json:    "linux-login"
linux-login-20140916.json:    "linux-login"
linux-login-20200525.json:    "linux-login"
linux-login-autoyast-20150724.json:    "linux-login"
linux-login-autoyast-20160822.json:    "linux-login"
linux-login-autoyast-20191126.json:    "linux-login",
linux-login-autoyast-20200203.json:    "linux-login"
linux-login-kubic-20180201.json:    "linux-login-microos"
linux-login-kubic-20190408.json:    "linux-login-microos"
linux-login-kubic-20190415.json:    "linux-login-microos"
linux-login-kubic-20191031.json:    "linux-login-microos"
linux-login-kubic-20200428.json:    "linux-login-microos"
linux-login-kubic-20200909.json:    "linux-login-microos"
```

Comment (jlausuch) [reply lib/microos.pm]: my bad, in Suse either. I'll try to fix that then adding that tag as well...

Comment (Vogtinator) [reply lib/microos.pm]: Thanks, then `linux-login-microos` can be replaced everywhere, which should make this simpler and more consistent as well. Might even make some needles redundant, resulting in quicker matching

**Comment by jlausuch** [comment]
I found [another problem](http://fromm.arch.suse.de/tests/489#step/health_check/22) in healt-check.
When calling `process_reboot` it will wait to match `grub2`  tag, but sometimes it's too slow and the grub2 is automatically accepted after 10s and it will show the login screen, so there is no need to wait for grub2 twice... 

Most of these "Stall" problems would be solved by disabling grub timeout in RAW images as we do for DVD flavor, but it's not possible as it's RO filesystem.

**Comment by Vogtinator** [comment]
> I found [another problem](http://fromm.arch.suse.de/tests/489#step/health_check/22) in healt-check.
> When calling `process_reboot` it will wait to match `grub2` tag, but sometimes it's too slow and the grub2 is automatically accepted after 10s and it will show the login screen, so there is no need to wait for grub2 twice...

IMO if openQA is too slow to match a 10s time window, something is really broken.

> Most of these "Stall" problems would be solved by disabling grub timeout in RAW images as we do for DVD flavor, but it's not possible as it's RO filesystem.

I'll copy what I wrote in the first comment:

> It is, just needs a new snapshot. I agree that it's cleaner to work without needing that though.

**Comment by jlausuch** [comment]
> IMO if openQA is too slow to match a 10s time window, something is really broken.

It is true. This is the reality, cause looking at the logs it's looking for a match among ~100 needles. I am trying to do some [cleanup](https://gitlab.suse.de/openqa/os-autoinst-needles-sles/-/merge_requests/1447) for SLE case, but I don't think it will help as there would be still tones of candidate needles, unless I unregister all needles except `ENV-MICROOS` tag and create all the installation ones with that tag... still, lots of needles without TAGs that would need a change... 

> > It is, just needs a new snapshot. I agree that it's cleaner to work without needing that though.
If you are not convinced about the changes in this PR, I will go this way... I'll create a module to do that small change in grub config and boot from it.

**Comment by Vogtinator** [comment]
>> IMO if openQA is too slow to match a 10s time window, something is really broken.
>
> It is true. This is the reality, cause looking at the logs it's looking for a match among ~100 needles. I am trying to do some cleanup for SLE case, but I don't think it will help as there would be still tones of candidate needles, unless I unregister all needles except ENV-MICROOS tag and create all the installation ones with that tag... still, lots of needles without TAGs that would need a change...

For openSUSE it looks like this: https://openqa.opensuse.org/tests/1440017#step/health_check/18
There are 52 needle canditates and it still matches with 9s left. Even with ~100 needles, that seems like something else is broken.

> If you are not convinced about the changes in this PR, I will go this way... I'll create a module to do that small change in grub config and boot from it.

It's ok, just the case with `linux-login-microos` vs. `linux-login` needed some cleanup, which seems to be done now.

The duplicate step with matching `linux-login`, then `linux-login-microos` will be fixed by adjusting the needles, so that would be covered as well in the future.

**Comment by jlausuch** [comment]
So, finally I decided to implement the disable grub approach. https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/11234

I put this PR as question mark, I think I prefer the other solution plus fixing some needles.
